File: frontend/src/types/instagramCheck.ts
import type { Source } from "./chat";

export type InstagramCheckStatus = "draft" | "evaluated";
export type InstagramClaimStatus = "active" | "removed";
export type InstagramTrafficLight = "green" | "yellow" | "red";
export type TaxPrepRiskLevel = "low" | "medium" | "high";
export type InstagramTipCategory =
  | "arbeitsmittel"
  | "homeoffice"
  | "pendeln"
  | "weiterbildung"
  | "sonderausgaben"
  | "vorsorge"
  | "haushaltsnahe_dienstleistungen"
  | "kinder/familie"
  | "kapital"
  | "betriebsausgaben"
  | "other";
export type InstagramReturnBucket =
  | "anlage_n"
  | "betriebsausgaben"
  | "sonderausgaben"
  | "vorsorge"
  | "haushaltsnahe_dienstleistungen"
  | "kinder/familie"
  | "kapital"
  | "other";

export interface InstagramImageRef {
  id: string;
  originalFilename: string;
  storedPath: string;
  contentType: string;
  sizeBytes: number;
}

export interface InstagramFollowUpQuestion {
  id: string;
  prompt: string;
  answer?: string | null;
}

export interface InstagramClaim {
  id: string;
  rawText: string;
  editedText: string;
  category: InstagramTipCategory;
  returnBucket: InstagramReturnBucket;
  status: InstagramClaimStatus;
  selectedForImport: boolean;
  followUpQuestions: InstagramFollowUpQuestion[];
}

export interface InstagramEvaluatedTip {
  claimId: string;
  title: string;
  normalizedTip: string;
  category: InstagramTipCategory;
  returnBucket: InstagramReturnBucket;
  trafficLight: InstagramTrafficLight;
  explanation: string;
  estimatedSavingEur?: number | null;
  requiredEvidence: string[];
  sources: Source[];
}

export interface InstagramPostCheck {
  sessionId: string;
  status: InstagramCheckStatus;
  images: InstagramImageRef[];
  claims: InstagramClaim[];
  evaluatedTips?: InstagramEvaluatedTip[] | null;
  updatedAt: Date;
}

export interface InstagramCheckSavePayload {
  claims: InstagramClaim[];
}

export interface TaxPrepItem {
  id: string;
  sessionId: string;
  source: "instagram_post";
  sourceClaimId: string;
  title: string;
  category: InstagramTipCategory;
  returnBucket: InstagramReturnBucket;
  estimatedSavingEur?: number | null;
  riskLevel: TaxPrepRiskLevel;
  requiredEvidence: string[];
  summary: string;
  status: "confirmed";
  createdAt: Date;
  updatedAt: Date;
}

File: backend/app/models/instagram_check.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.idea_transfer import IdeaTransferSource


InstagramCheckStatus = Literal["draft", "evaluated"]
ClaimStatus = Literal["active", "removed"]
TrafficLight = Literal["green", "yellow", "red"]
RiskLevel = Literal["low", "medium", "high"]
TipCategory = Literal[
    "arbeitsmittel",
    "homeoffice",
    "pendeln",
    "weiterbildung",
    "sonderausgaben",
    "vorsorge",
    "haushaltsnahe_dienstleistungen",
    "kinder/familie",
    "kapital",
    "betriebsausgaben",
    "other",
]
ReturnBucket = Literal[
    "anlage_n",
    "betriebsausgaben",
    "sonderausgaben",
    "vorsorge",
    "haushaltsnahe_dienstleistungen",
    "kinder/familie",
    "kapital",
    "other",
]
TaxPrepSource = Literal["instagram_post"]
TaxPrepStatus = Literal["confirmed"]


class InstagramImageRef(BaseModel):
    id: str
    original_filename: str
    stored_path: str
    content_type: str
    size_bytes: int


class InstagramFollowUpQuestion(BaseModel):
    id: str
    prompt: str = Field(..., min_length=1, max_length=400)
    answer: str | None = Field(default=None, max_length=1000)


class InstagramClaim(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    raw_text: str = Field(..., min_length=1, max_length=1000)
    edited_text: str = Field(..., min_length=1, max_length=1000)
    category: TipCategory
    return_bucket: ReturnBucket
    status: ClaimStatus = "active"
    selected_for_import: bool = False
    follow_up_questions: list[InstagramFollowUpQuestion]


class InstagramEvaluatedTip(BaseModel):
    model_config = ConfigDict(extra="forbid")

    claim_id: str
    title: str = Field(..., min_length=1, max_length=200)
    normalized_tip: str = Field(..., min_length=1, max_length=1200)
    category: TipCategory
    return_bucket: ReturnBucket
    traffic_light: TrafficLight
    explanation: str = Field(..., min_length=1, max_length=2000)
    estimated_saving_eur: int | None = Field(default=None, ge=0)
    required_evidence: list[str]
    sources: list[IdeaTransferSource]


class InstagramPostCheck(BaseModel):
    session_id: str
    status: InstagramCheckStatus
    images: list[InstagramImageRef]
    claims: list[InstagramClaim]
    evaluated_tips: list[InstagramEvaluatedTip] | None = None
    updated_at: datetime


class InstagramClaimUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    edited_text: str = Field(..., min_length=1, max_length=1000)
    category: TipCategory
    return_bucket: ReturnBucket
    status: ClaimStatus
    selected_for_import: bool = False
    follow_up_questions: list[InstagramFollowUpQuestion]


class InstagramCheckSaveRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    claims: list[InstagramClaimUpdate]


class InstagramCheckEvaluateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    session_id: str
    tax_year: int = 2025


class TaxPrepItem(BaseModel):
    id: str
    session_id: str
    source: TaxPrepSource
    source_claim_id: str
    title: str
    category: TipCategory
    return_bucket: ReturnBucket
    estimated_saving_eur: int | None = None
    risk_level: RiskLevel
    required_evidence: list[str]
    summary: str
    status: TaxPrepStatus
    created_at: datetime
    updated_at: datetime
